implementation/discovery/etcd/src/main/java/org/inaetics/dronessimulator/discovery/etcd/EtcdConfigDiscoverer.java

```diff
@@ -0,0 +1,123 @@
+package org.inaetics.dronessimulator.discovery.etcd;
+
+import org.apache.log4j.Logger;
+import org.osgi.service.cm.Configuration;
+import org.osgi.service.cm.ConfigurationAdmin;
+
+import java.io.IOException;
+import java.util.*;
+
+/**
+ * Discovers configurations and publishes these as services.
+ */
+public class EtcdConfigDiscoverer implements Runnable {
+    private static final Logger logger = Logger.getLogger(EtcdConfigDiscoverer.class);
+
+    /** The discoverer used to find configs. */
+    private EtcdDiscoverer discoverer;
+
+    /** The configuration admin used to register config services. */
+    private ConfigurationAdmin configurationAdmin;
+
+    private HashMap<String, String> configs = new HashMap<>();
+
+    /**
+     * Instantiates a new config discoverer.
+     * @param discoverer The etcd discoverer to use to find configs.
+     */
+    public EtcdConfigDiscoverer(EtcdDiscoverer discoverer, ConfigurationAdmin configurationAdmin) {
+        this.discoverer = discoverer;
+        this.configurationAdmin = configurationAdmin;
+
+        // Initialize some fields
+        this.configs = new HashMap<>();
+    }
+
+    @Override
+    public void run() {
+        logger.info("Started etcd config discoverer");
+
+        Collection<String> currentPaths = this.configs.keySet();
+
+        while(!Thread.interrupted()) {
+            // Wait for update in directory
+            Collection<String> collectedPaths = this.discoverer.getDiscoverableConfigs(true);
+            Collection<String> newPaths = new HashSet<>();
+            Collection<String> stalePaths = new HashSet<>(currentPaths);
+
+            logger.debug("Found {} configs in etcd", collectedPaths.size());
+
+            // Process paths
+            for (String path : collectedPaths) {
+                if (this.configs.containsKey(path)) {
+                    logger.debug("Config path {} is still there", path);
+                    stalePaths.remove(path);
+                } else {
+                    logger.debug("Config path {} is new", path);
+                    newPaths.add(path);
+                }
+            }
+
+            newPaths.forEach(this::registerInstance);
+            stalePaths.forEach(this::unregisterInstance);
+
+            // Update paths
+            currentPaths = this.configs.keySet();
+        }
+
+        logger.info("Stopping etcd config discoverer");
+    }
+
+    /**
+     * Registers a new discovered config with properties located on the given path.
+     * @param instancePath The instance path for the config.
+     */
+    private void registerInstance(String instancePath) {
+        logger.debug("Registering {} as a discovered config", instancePath);
+
+        // Split path
+        String[] pathSegments = EtcdDiscoverer.splitInstancePath(instancePath);
+        String type = pathSegments[0];
+        String group = pathSegments[1];
+        String name = pathSegments[2];
```

Comment:
Idee om eerst pathSegments.length te checken en als deze !=3 een exceptie te gooien? (Of i.i.g. te loggen?)